build.py

```diff
@@ -0,0 +1,228 @@
+        help='Clean the build targets, and remove build and intermediate files')
+    arg_parser.add_argument(
+        '--skip-update-deps', dest='skip_update_deps',
+        action='store_true', default=False,
+        help='Skip updating external dependencies')
+    arg_parser.add_argument(
+        '--skip-code-style', dest='skip_code_style',
+        action='store_true', default=False,
+        help='Skip applying C++ code style before compiling')
+    arg_parser.add_argument(
+        '--static-analysis', dest='static_analysis',
+        action='store_true', default=False,
+        help='Run static analysis on the code')
+    return arg_parser.parse_args()
+
+
+def update_external_dependencies(args):
+    '''
+    Update git submodules
+    '''
+    if not args.skip_update_deps:
+        update_git_submodule_result = subprocess.run(
+            ['git', 'submodule', 'update', '--init'], cwd=BUILD_ROOT)
+        if 0 != update_git_submodule_result.returncode:
+            raise BuildError('failed to update git submodules')
+
+
+def build_dir(args):
+    '''
+    Get the CMake build directory
+    '''
+    return os.path.join(BUILD_ROOT,
+                        BUILD_CONFIGS[args.configuration],
+                        platform.system().lower(),
+                        args.architecture)
+
+
+def cmake_version():
+    '''
+    Get the CMake version
+    '''
+    cmake_version_output = subprocess.check_output(
+        ['cmake', '--version']).decode('utf-8')
+    match = re.match(
+        r'cmake version (?P<version>[\d\.]+)', cmake_version_output)
+    if match is None:
+        raise BuildError('failed to get CMake version')
+    cmake_version = distutils.version.StrictVersion(match.group('version'))
+    return cmake_version
+
+
+def cmake_generate_options(args):
+    '''
+    CMake build file generation options
+    '''
+    generate_options = []
+    if args.clean or args.clobber:
+        generate_options.append('-DAPPLY_CPP_CODE_STYLE=OFF')
+        generate_options.append('-DRUN_STATIC_ANALYSIS=OFF')
+    else:
+        if args.skip_code_style:
+            generate_options.append('-DAPPLY_CPP_CODE_STYLE=OFF')
+        else:
+            generate_options.append('-DAPPLY_CPP_CODE_STYLE=ON')
+        if args.static_analysis:
+            generate_options.append('-DRUN_STATIC_ANALYSIS=ON')
+        else:
+            generate_options.append('-DRUN_STATIC_ANALYSIS=OFF')
+    return generate_options
+
+
+def cmake_generate_build_files(args):
+    '''
+    Use CMake to generate build files
+    '''
+    system = platform.system().lower()
+    cmake_generate_args = [
+        'cmake',
+        '-DCMAKE_INSTALL_PREFIX=' + os.path.join(build_dir(args), 'install')]
+    cmake_generate_env = os.environ.copy()
+    if 'windows' == system:
+        if 'x64' == args.architecture:
+            cmake_generate_args.extend(['-A', 'x64'])
+    else:
+        if 'debug' == args.configuration:
```

Comment:
Elsewhere, `args.configuration.capitalize()` is used to derive the value of `CMAKE_BUILD_TYPE`.  Please do so here for consistency; it'll also be shorter code that doesn't require an `if/else':
```
cmake_generate_args.append('-DCMAKE_BUILD_TYPE={}'.format(args.configuration.capitalize())
```